### PR TITLE
RTEMIS-198: Added Student Tracking Dashboard for School Admin.

### PR DESCRIPTION
**Pull request by RohanSingh-inno**
Ticket # : https://innoraft.atlassian.net/browse/RTEMIS-198

---
### What the PR does
`(Explain fixes, changes and impacts)`
-  Added a view.
-  Added a menu link for the same.
-  Modified user roles permission.
-  Added a form mode.

### CODE REVIEW

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module [hunk 1]
```diff
@@ -24,3 +28,185 @@ function rte_mis_student_tracking_get_allowed_values(FieldStorageDefinitionInter
...
+    $form['total_students'] = [
+      '#type' => 'markup',
+      '#markup' => '<ul><li>' . implode('</li><li>', $items) . '</li></ul>',
+    ];
+
+    $form['#cache'] = [
+      'tags' => [
+        'config:rte_mis_student_tracking.settings',
+        'config:rte_mis_school.settings',
+      ],
```
Comment (kunal-sin) [inline modules/rte_mis_student_tracking/rte_mis_student_tracking.module:64]: Please fetch the cache tag from config object.

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module [hunk 2]
```diff
@@ -24,3 +28,185 @@ function rte_mis_student_tracking_get_allowed_values(FieldStorageDefinitionInter
...
+ * Function to get the options.
+ */
+function rte_mis_student_tracking_class_list_options() {
+  $options = [];
+  $tracking_config = \Drupal::config('rte_mis_student_tracking.settings') ?? NULL;
+  $school_config = \Drupal::config('rte_mis_school.settings') ?? NULL;
+  $selected_class = $tracking_config->get('allowed_class_list');
+  $entry_class = $school_config->get('field_default_options.class_level') ?? [];
+  sort($selected_class);
+  foreach ($selected_class as $class) {
```
Comment (kunal-sin) [inline modules/rte_mis_student_tracking/rte_mis_student_tracking.module:79]: Please add some comment here to explain the reason.

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module [hunk 3]
```diff
@@ -24,3 +28,185 @@ function rte_mis_student_tracking_get_allowed_values(FieldStorageDefinitionInter
...
+      $student_count = rte_mis_student_tracking_get_student_count($class);
+      $options[$class] = '<h2>' . t('@class', ['@class' => $entry_class[$class]]) . '</h2><br>' . t('@count', ['@count' => $student_count]);
+    }
+  }
+  return $options;
+}
+
+/**
+ * Get the count of students based on class.
+ */
```
Comment (kunal-sin) [inline modules/rte_mis_student_tracking/rte_mis_student_tracking.module:90]: Function param missing

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module [hunk 4]
```diff
@@ -24,3 +28,185 @@ function rte_mis_student_tracking_get_allowed_values(FieldStorageDefinitionInter
...
+    // Get the current loggedIn user.
+    $currentUser = \Drupal::currentUser();
+    if (array_intersect(['school_admin'], $currentUser->getRoles())) {
+      // Load the user entity.
+      $user = \Drupal::entityTypeManager()->getStorage('user')->load($currentUser->id() ?? NULL);
+      // Check if the user is an instance of userInterface.
+      if ($user instanceof UserInterface) {
+        // Get the school id linked with the current loggedIn user.
+        $schoolId = $user->get('field_school_details')->getString() ?? NULL;
+        // dd($schoolId);
```
Comment (kunal-sin) [inline modules/rte_mis_student_tracking/rte_mis_student_tracking.module:114]: Please remove the commented code.

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module [hunk 5]
```diff
@@ -24,3 +28,185 @@ function rte_mis_student_tracking_get_allowed_values(FieldStorageDefinitionInter
...
+  if ($view->id() == 'student_tracking_dashboard') {
+    // Get the current loggedIn user.
+    $currentUser = \Drupal::currentUser();
+    if (array_intersect(['school_admin'], $currentUser->getRoles())) {
+      // Load the user entity.
+      $user = \Drupal::entityTypeManager()->getStorage('user')->load($currentUser->id() ?? NULL);
+      // Check if the user is an instance of userInterface.
+      if ($user instanceof UserInterface) {
+        // Get the school id linked with the current loggedIn user.
+        $schoolId = $user->get('field_school_details')->getString() ?? NULL;
```
Comment (kunal-sin) [inline modules/rte_mis_student_tracking/rte_mis_student_tracking.module:113]: `$school_id`

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module [hunk 6]
```diff
@@ -24,3 +28,185 @@ function rte_mis_student_tracking_get_allowed_values(FieldStorageDefinitionInter
...
+  foreach ($selected_class as $class) {
+    if (isset($entry_class[$class])) {
+      $total_count[$class] = 0;
+    }
+  }
+  $currentUser = \Drupal::currentUser();
+  // Get the current loggedIn User.
+  $currentUserId = $currentUser->id();
+  // Get the user roles.
+  $currentUserRoles = $currentUser->getRoles(TRUE);
```
Comment (kunal-sin) [inline modules/rte_mis_student_tracking/rte_mis_student_tracking.module:193]: `$current_user_roles`

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module [hunk 7]
```diff
@@ -24,3 +28,185 @@ function rte_mis_student_tracking_get_allowed_values(FieldStorageDefinitionInter
...
+  if (in_array('school_admin', $currentUserRoles)) {
+    // Get the school linked to the current user.
+    $currentUserSchool = \Drupal::entityTypeManager()->getStorage('user')->load($currentUserId)->get('field_school_details')->getString() ?? NULL;
+    // Check within all the schools where total class is defined.
+    $educationDetails = \Drupal::entityTypeManager()->getStorage('mini_node')->load($currentUserSchool)->get('field_education_details') ?? NULL;
```
Comment (kunal-sin) [inline modules/rte_mis_student_tracking/rte_mis_student_tracking.module:198]: `$education_details`

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module [hunk 8]
```diff
@@ -24,3 +28,185 @@ function rte_mis_student_tracking_get_allowed_values(FieldStorageDefinitionInter
...
+    $educationDetailsEntity = $educationDetails ? $educationDetails->referencedEntities() : NULL;
```
Comment (kunal-sin) [inline modules/rte_mis_student_tracking/rte_mis_student_tracking.module:199]: `$education_details_entity`

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module [hunk 9]
```diff
@@ -24,3 +28,185 @@ function rte_mis_student_tracking_get_allowed_values(FieldStorageDefinitionInter
...
+    foreach ($educationDetailsEntity as $value) {
+      // Fee Details for each education detail.
+      $feeDetails = $value->get('field_fee_details')->referencedEntities();
```
Comment (kunal-sin) [inline modules/rte_mis_student_tracking/rte_mis_student_tracking.module:202]: `$fee_details`

File: config/install/views.view.student_tracking_dashboard.yml [hunk 10]
```diff
@@ -0,0 +1,869 @@
...
+          element_label_colon: true
+          element_wrapper_type: ''
+          element_wrapper_class: ''
+          element_default_classes: true
+          empty: ''
+          hide_empty: false
+          empty_zero: false
+          hide_alter_empty: false
+      pager:
+        type: mini
```
Comment (kunal-sin) [inline config/install/views.view.student_tracking_dashboard.yml:593]: Please use full pager.

Comment (RohanSingh-inno) [reply config/install/views.view.student_tracking_dashboard.yml]: ok

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module (hunk 9 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student_tracking/rte_mis_student_tracking.module]: ok

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module (hunk 8 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student_tracking/rte_mis_student_tracking.module]: ok

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module (hunk 7 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student_tracking/rte_mis_student_tracking.module]: ok

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module (hunk 6 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student_tracking/rte_mis_student_tracking.module]: ok

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module (hunk 5 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student_tracking/rte_mis_student_tracking.module]: ok

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module (hunk 4 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student_tracking/rte_mis_student_tracking.module]: ok

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module (hunk 3 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student_tracking/rte_mis_student_tracking.module]: ok

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module (hunk 2 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student_tracking/rte_mis_student_tracking.module]: ok

File: modules/rte_mis_student_tracking/rte_mis_student_tracking.module (hunk 1 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student_tracking/rte_mis_student_tracking.module]: ok

**Pull request merged**